### PR TITLE
Beast Wizard:  Helpers to validate mappings and generate TMCF

### PR DESCRIPTION
**Pull request by pradh**
Please have a look at the mapping validation checks and see if they make sense.  A new thing is that under if COLUMN_HEADER type is never provided, then the value column must be explicitly identified.

### CODE REVIEW

**Review by pradh — COMMENTED** [review]
Thanks for the review!

File: wizard/src/utils/tmcf_generation.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  // Do one pass over the mappings building the common PVs in all TMCF nodes.
+  // Everything other than COLUMN_HEADER mappings get repeated in every node.
+  mappings.forEach((mval: MappingVal, mthing: MappedThing) => {
+    const mappedProp = MAPPED_THING_TO_SVOBS_PROP.get(mthing);
+    if (mval.type == MappingType.CONSTANT) {
+      // Constants are references except when it is a date.
+      commonPVs.push(getConstPV(mappedProp, mval.constant));
+    } else if (mval.type == MappingType.COLUMN) {
+      if (mthing == MappedThing.PLACE) {
+        if (mval.placeProperty == DCID_PROP) {
```
[the remark this answers is not in the document]
Comment (jehangiramjad) [reply wizard/src/utils/tmcf_generation.ts:89]: I am not the expert here but shouldn't that also be the case?

File: wizard/src/utils/validation.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+ * (3) At most one MappedThing can be a COLUMN_HEADER
+ * (4) If there are no COLUMN_HEADER mappings, VALUE should be mapped
+ *     (a) Conversely, if COLUMN_HEADER mapping exists, VALUE must not be mapped
+ * (5) At least one of the mappings must specify a COLUMN or a COLUMN_HEADER
+ * (6) PLACE, when mapped as a COLUMN, must specify placeProperty
+ * (7) VALUE, if it appears, has to be COLUMN
+ * TODO: Allow placeProperty for all mapping types of PLACE
+ */
+export function checkMappings(mappings: Mapping): Array<string> {
+  const errors = Array<string>();
```
[the remark this answers is not in the document]
Comment (jehangiramjad) [reply wizard/src/utils/validation.ts:40]: Great, thanks!

File: wizard/src/utils/tmcf_generation.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pradh) [reply wizard/src/utils/tmcf_generation.ts:89]: That's covered by the else case? Where we will end up something like:

Node: E:CSV->E0
typeOf: dcs:Country
name: C:CSV->name_col

Node: E:CSV->E1
typeOf: dcs:StatVarObservation
observationAbout: E:CSV->E0

[the remark this answers is not in the document]
Comment (jehangiramjad) [reply wizard/src/utils/tmcf_generation.ts:89]: Ah ok!

File: wizard/src/utils/tmcf_generation.ts [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  [MappedThing.VALUE, VALUE_PROP],
+]);
+
+//
+// A set of node/pv generation helpers
+//
+
+function initNode(idx: number, type: string): Array<string> {
+  const pvs = Array<string>();
+  pvs.push("Node: E:" + FIXED_CSV_TABLE + "->E" + idx.toString());
```
Comment (chejennifer) [inline wizard/src/utils/tmcf_generation.ts:44]: maybe we can use template strings here and elsewhere where we are concatenating strings: 
```
'Node: E:${FIXED_CSV_TABLE}->E${idx.toString()}`
```

File: wizard/src/utils/validation.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    if (!mappings.has(mthing)) {
+      // Check #2
+      errors.push("Missing required mapping for " + mthing);
+    }
+  }
+
+  const colHdrThings = Array<string>();
+  let numNonConsts = 0;
+  mappings.forEach((mval: MappingVal, mthing: MappedThing) => {
+    if (mval.type == MappingType.COLUMN) {
```
Comment (chejennifer) [inline wizard/src/utils/validation.ts:55]: in javascript (/typescript), we usually want to use "===" instead of "=="

**Review by chejennifer — COMMENTED** [review]
nice!

File: wizard/src/utils/tmcf_generation.ts (hunk 3 above)
Comment (pradh) [reply wizard/src/utils/tmcf_generation.ts]: Very nice!

File: wizard/src/utils/validation.ts (hunk 4 above)
Comment (pradh) [reply wizard/src/utils/validation.ts]: Ah, thanks, I forgot...

**Review by pradh — COMMENTED** [review]
Thanks for the review!

I've also added validation/tmcf support for PLACE with placeType occurring in both COLUMN and COLUMN_HEADER mappings.

File: wizard/src/utils/tmcf_generation.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (pradh) [reply wizard/src/utils/tmcf_generation.ts]: I've added a test case for this.

**Review by jehangiramjad — APPROVED** [review]
Thanks!

**Review by chejennifer — APPROVED** [review]
Thanks!

**Comment by pradh** [comment]
Thanks for the reviews!

**Pull request merged**